feat!: refactor types to work better with typescript 4.7

Julusian opened the pull request:
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

feature/fix

* **What is the current behavior?** (You can also link to an open issue here)

The typings are hard to use with typescript 4.7+, as some of the type unions we perform in blueprints and core produce an error `Type instantiation is excessively deep and possibly infinite.`.

This happens because of how the types we export (eg `TimelineObjCCGRoute`) replace the `content` type of `TSRTimelineObj`, so when we perform a union with where it is present, the compiler gets confused.

Minimal demo: https://www.typescriptlang.org/play?ts=4.7.4#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKihSIBPBkNADeoAAikDrqkD6Q-DihMBSckADy0gBWAMKFAOIAShy26cjQWZCF8JwKTSmFDIIEjoXMAL6gAGZsiKAA5KgZtdkAtIKoDApTToLw0HpkU6iE+iMhYRHR3hN1efmQsKgNTgCiAB6ofYNIowACy-0UkFMM7OiI8xSNYDSaDsSAEShNQRTLIKVj-Ro7RQITizMQALkONWOBWK5UqKAAZJjMtkTmcLhFILcaABeUBRRQASAo3AxIxGvCZkE4DGBkAxMVmjlQGIADKAepzGeCATY8BiAExSxh4SBkNm82AcpnI5oi+lMxncBJJFIY+KJCwpAKFACCxFotrKpSljM2RAx1RJ9Uaeta7U6ZG6ATKOQAqjgrq6-gQiNwADJMNVs439DmgFTETDwCSgABi7Gg0CZksUPSAA

* **What is the new behavior (if this is a feature change)?**

As all the objects are identical except for the `content`, `TSRTimelineObj` is now a generic, taking the type of `Content` as the first parameter. The `TimelineObjCCGRoute` and other types have been changed to `TimelineContentCCGRoute`, and are now to be used as `TSRTimelineObj<TimelineContentCCGRoute>`.

`TSRTimelineObj<TSRTimelineContent>` can be used when the content is unknown but must be one of the ones we have defined.

This means that in sofie blueprints/core, instead of the problematic `TimelineObjCCGRoute & TimelineObjectCoreExt`, you can do `TimelineObjectCoreExt<TimelineContentCCGRoute>`

* **Other information**:

This did result in ~400 type errors in the NRK blueprints, but it was fairly straightforward to fix it all up. Some bits were manual, other bits were possible to do with regexes. It took about an hour to resolve.

For those reviewing, make sure to enable the 'No whitespace' option in github as that halves the number of lines with changes. Changing the types from `TimelineObjCCGRoute` to `TimelineContentCCGRoute` resulted in all of the lines inside `content` to change indentation by one level.

The change to typescript version should be reverted before merging, it is there to ensure it works as expected.

Sofie core PR: https://github.com/nrkno/sofie-core/pull/793
NRK blueprints: https://github.com/nrkno/sofie-nrk-blueprints/pull/205

baltedewit reviewed (approved):
I wanted to do this change earlier for stronger typings so overall I'm really positive about this 👍

Julusian commented on packages/timeline-state-resolver-types/src/casparcg.ts:
```diff
@@ -59,7 +59,7 @@ export interface TimelineStingTransition extends TimelineTransitionBase {
 }
 
-export interface TimelineObjCCGProducerContentBase {
+export interface TimelineContentCCGProducerContentBase {
```
can you tell I did some search and replace and didnt read through every rename? ;)

Julusian commented on packages/timeline-state-resolver-types/src/superfly-timeline.ts:
```diff
@@ -27,25 +27,23 @@ export interface ResolveOptions {
 	layer: string | number
 	/** Group children */
-	children?: Array<TimelineObject>
+	children?: Array<TimelineObject<any>>
```
I have a feeling I had an issue with `unknown` causing something to not work. Changing it now doesnt have any issue (unless the error was from blueprints/core), so lets start with `unknown`

Julusian replied:
I found it. A load of the tests in `superfly-timeline.spec.ts` start failing if I change that.

I think that to fix this will require some additional larger changes to types, potentially including making supertimeline itself generic. As it's not an quick fix, I don't think I should spend more time on this.

Julusian commented:
@baltedewit I've fixed up those casts you raised, but they are present in most integrations and some are harder to fix than others so some have been skipped.
To do so I changed from using the superfly-timeline version of `TimelineState` to the one from tsr-types.

Much more can be done, especially around `Mapping`, but that becomes a larger task of strengthening the typings in this library. Hopefully this has done enough groundwork for that to be continued as other changes are made here
